diffedit.md

```diff
@@ -0,0 +1,752 @@
+
+
+```python
+# denoising loop
+for i, t in enumerate(tqdm(denoising_scheduler.timesteps[start_step:])):
+    current_step = i + start_step
+    next_step = current_step + 1
+
+    input = torch.cat([noisy_latents] * 2)
+    scaled_input = denoising_scheduler.scale_model_input(input, t)
+
+    # predict the noise residual
+    with torch.no_grad(): 
+        pred = unet(scaled_input, t, encoder_hidden_states=text_embeddings).sample
+
+    # perform guidance
+    pred_uncond, pred_text = pred.chunk(2)
+    pred_nonscaled = pred_uncond + guidance_scale * (pred_text - pred_uncond)
+    pred = pred_nonscaled
+
+    # compute the "previous" noisy sample
+    if (next_step < num_inference_steps):
+        xt = create_noise(original_image_scheduler, original_image_latents, start_step=next_step)
+    else:
+        xt = original_image_latents
+
+    yt = denoising_scheduler.step(pred, t, noisy_latents).prev_sample
+
+    noisy_latents = mask*yt+((1-mask)*xt)
+
+    if callback:
+        callback(noisy_latents)
+final_latents = noisy_latents
+```
+
+
+      0%|          | 0/38 [00:00<?, ?it/s]
+
+
+Once the denoising loop is complete, we can decode the final latents to see what our generated "zebra" DiffEdit image looks like!
+
+
+```python
+images_to_grid([original_image, to_image(final_latents)], 2)
+```
+
+
+    
+![png](diffedit_files/diffedit_83_0.png)
+    
+
+
+And there we have it! Our original horse has been completely replaced with our generated zebra image, and through the power of the DiffEdit process, you can see that the background has been untouched outside of the  masked area.
+
+We also added a callback so that we could see the denoising process as it progressed. We started at step 12 since we wanted the original image to still guide the pose of our zebra in our output image.
+
+
+```python
+images_indexes = [(to_image(l), i) for i, l in list(enumerate(itermediate_latents, start_step))[::5]]
+i, l = zip(*images_indexes)
+labels = [f"step {label}" for label in l]
+images_to_grid([*i, to_image(final_latents)], 3, titles=[*labels, f"step {num_inference_steps}"])
+```
+
+
+    
+![png](diffedit_files/diffedit_86_0.png)
+    
+
+
+## Putting it all together in functions
```

Comment:
As this is already a rather long (but thoroughly enjoyable!) post. I'd recommend dropping this section. You can always link to a repo with the full code in it.